### PR TITLE
kafka exporter: seperate message conversion for otlp and jaeger

### PR DESCRIPTION
**Pull request by Frefreak**
**Description:** As per the issue linked below, for kafka exporter currently when converting messages to sarama `ProducerMessage`, we uses the same function which sets key with messges's `Key` field. However for **otlp_proto**, unlike **jaeger_json** or **jaeger_proto** the `Key` is not set and thus always an empty byte array, resulting in imbalanced partition write. This commit generalize the message conversion as an interface and thus separate the conversion part between **otlp_proto** and **jaeger_json**/**jaeger_proto** (so using different logic without a runtime `if` check).

**Link to tracking Issue:** #3162 
**Testing:** tested performance with/without this commit.
v0.24.0: \~300K span/s (sent_span)
v0.26.0: 50\~60K span/s (after the commit that introduced this `Key` things for jaeger)
after this commit: \~300K span/s

Also `go test`-ed in kafkaexporter folder.

Please see issue for details.

### CODE REVIEW

File: exporter/kafkaexporter/converter.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// limitations under the License.
+
+package kafkaexporter
+
+import (
+	"github.com/Shopify/sarama"
+)
+
+// MessageConverter converts our message type to sarama message type
+type MessageConverter interface {
```
Comment (bogdandrutu) [inline exporter/kafkaexporter/converter.go:22]: I think we can make this better by changing the "marshaler" interface to produce `[]*sarama.ProducerMessage` directly? What do you think? In that case we skip the `Message` creation (unnecessary allocation for the slice of messages).

File: exporter/kafkaexporter/converter.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+// ConvertMessages converter for otlp_proto
+func (mp *OTLPProtoConverter) ConvertMessages(messages []Message, topic string) []*sarama.ProducerMessage {
+	producerMessages := make([]*sarama.ProducerMessage, len(messages))
+	for i := range messages {
+		producerMessages[i] = &sarama.ProducerMessage{
+			Topic: topic,
+			Value: sarama.ByteEncoder(messages[i].Value),
+		}
+	}
+	return producerMessages
```
Comment (bogdandrutu) [inline exporter/kafkaexporter/converter.go:43]: Not sure about the sarama behavior, but you can give a try to reduce allocations here:
```
	producerMessagesOrigs := make([]sarama.ProducerMessage, len(messages))
	producerMessages := make([]*sarama.ProducerMessage, len(messages))
	for i := range messages {
		producerMessages[i] = &producerMessagesOrigs[i]
		producerMessages[i].Topic = topic
		producerMessages[i].Value = sarama.ByteEncoder(messages[i].Value),
	}
	return producerMessages
```

File: exporter/kafkaexporter/converter.go (hunk 1 above)
Comment (Frefreak) [reply exporter/kafkaexporter/converter.go:22]: > I think we can make this better by changing the "marshaler" interface to produce `[]*sarama.ProducerMessage` directly? What do you think? In that case we skip the `Message` creation (unnecessary allocation for the slice of messages).

During my implementation it did feels kind of tedious to create another "converter" interface to adjust such a simple behavior. I agree that changing the marshaler interface to generate `samara.ProducerMessage` directly seems like a better idea. I will make the modification accordingly.

File: exporter/kafkaexporter/converter.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// Encoding return encoding for jaeger_proto
+func (mp *JaegerJSONConverter) Encoding() string {
+	return "jaeger_json"
+}
+
+// getConverters returns all pre-configured converters for supported encodings
+func getConverters() map[string]MessageConverter {
```
Comment (rakyll) [inline exporter/kafkaexporter/converter.go:96]: Maybe call this something like defaultConverters.

File: exporter/kafkaexporter/converter.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	otlppb := &OTLPProtoConverter{}
+	jaegerProto := &JaegerProtoConverter{}
+	jaegerJSON := &JaegerJSONConverter{}
+
+	return map[string]MessageConverter{
```
Comment (rakyll) [inline exporter/kafkaexporter/converter.go:101]: Put things inline:

return map[string]MessageConverter{
                 otlppb.Encoding():      &OTLPProtoConverter{},
		jaegerProto.Encoding(): &JaegerProtoConverter{},
		jaegerJSON.Encoding():  &JaegerJSONConverter{},
}

Avoid variables to reduce verbosity.

**Review by rakyll — CHANGES_REQUESTED** [review]

File: exporter/kafkaexporter/converter.go (hunk 4 above)
Comment (Frefreak) [reply exporter/kafkaexporter/converter.go]: Hi, as per the previous discussion I modified marshaler to directly convert pdata.traces/logs/metrics to sarama.ProducerMessage now, so no converter is needed now.

**Comment by Frefreak** [comment]
Updated. Now marshal transform `Traces`/`Metrics`/`Logs` to `sarama.ProducerMessage` directly instead of going through an inter-medium "Message" struct type.

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Lint errors:

```
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
golangci-lint run --allow-parallel-runners
exporter/kafkaexporter/factory_test.go: Expected no more than 3 groups, got 4
exporter/kafkaexporter/jaeger_marshaler.go: Expected no more than 3 groups, got 4
exporter/kafkaexporter/jaeger_marshaler_test.go: Expected no more than 3 groups, got 4
```

Also please add an entry in the CHANGELOG since you still need to update the PR

**Comment by jpkrohling** [comment]
Approving workflows to be executed.

**Pull request merged**